bns-test/test/name_register.test.ts

```diff
@@ -0,0 +1,345 @@
+          cases[0].namespace,
+          "Bob",
+          cases[0].salt, 
+          2560000, { sender: cases[0].nameOwner });
+        expect(receipt.success).eq(true);
+        expect(receipt.result).include('u44');
+      // });
+  
+      // it("Bob registering the name 'Bob.blockstack' should fail", async () => {
+        receipt = await bns.nameRegister(
+          cases[0].namespace, 
+          "Bob", 
+          cases[0].salt, 
+          cases[0].zonefile, { sender: cases[0].nameOwner });
+        expect(receipt.error).include('2022');
+        expect(receipt.success).eq(false);
+    //   });
+    // });
+
+    // describe("Given an existing pre-order of the name 'bob.blockstack', initiated by Bob at block #21", () => {
+
+    //   beforeEach(async () => {
+        receipt = await bns.namePreorder(
+          cases[0].namespace,
+          "bob",
+          cases[0].salt, 
+          2560000, { sender: cases[0].nameOwner });
+        expect(receipt.success).eq(true);
+        expect(receipt.result).include('u46');
+      // });
+  
+      // describe("Bob registering the name 'bob.blockstack'", async () => {
+
+      //   it("should succeed", async () => {
+          receipt = await bns.nameRegister(
+            cases[0].namespace, 
+            "bob", 
+            cases[0].salt, 
+            cases[0].zonefile, { sender: cases[0].nameOwner });
+          expect(receipt.result).include('true');
+          expect(receipt.success).eq(true);
+
+          receipt = await bns.getNameZonefile(
+            cases[0].namespace, 
+            "bob", { sender: cases[0].nameOwner });
+          expect(receipt.result).include('0x30303030');
+          expect(receipt.success).eq(true);
+        // });
+
+        // it("should fail registering twice", async () => {
+          receipt = await bns.nameRegister(
+            cases[0].namespace, 
+            "bob", 
+            cases[0].salt, 
+            cases[0].zonefile, { sender: cases[0].nameOwner });
+          expect(receipt.error).include('2004');
+          expect(receipt.success).eq(false);
+      //   });
+
+      // });
+
+      // describe("Charlie registering 'bob.blockstack'", () => {
+
+      //   it("should fail", async () => {
+
+          receipt = await bns.namePreorder(
+            cases[0].namespace,
+            "bob",
+            cases[0].salt, 
+            2560000, { sender: charlie });
+          expect(receipt.success).eq(true);
+          expect(receipt.result).include('u50');
+  
+          receipt = await bns.nameRegister(
+            cases[0].namespace, 
+            "bob", 
+            cases[0].salt, 
+            cases[0].zonefile, { sender: charlie });
+          expect(receipt.error).include('2004');
+          expect(receipt.success).eq(false);
+      //   });
+      // });
+
+      // // todo(ludo): investigate the case where 'name.namespace' was preordered while being unavailable
+      // // and then became available
```

Comment:
Are you asking what happens if `name.namespace` is preordered, the namespace is launched, and then `name.namespace` is registered?  I could go either way on this, but there should be a test case that demonstrates the rule.

Reply:
EDIT: I see from the code that the preorder has to have been sent after the namespace was launched.